dbms/src/DataStreams/SquashingHashJoinBlockTransform.cpp

```diff
@@ -0,0 +1,98 @@
+// Copyright 2022 PingCAP, Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <DataStreams/SquashingHashJoinBlockTransform.h>
+
+namespace DB
+{
+
+SquashingHashJoinBlockTransform::SquashingHashJoinBlockTransform(UInt64 max_block_size_)
+    : output_rows(0)
+    , max_block_size(max_block_size_)
+    , join_finished(false)
+    , over_limit(false)
+{}
+
+void SquashingHashJoinBlockTransform::handleOverLimitBlock()
+{
+    // if over_limit_block is not null, we need to push it into blocks first.
+    if (over_limit_block.has_value())
+    {
+        output_rows += over_limit_block->rows();
+        blocks.push_back(std::move(over_limit_block.value()));
+        over_limit_block = {};
+    }
+}
+
+void SquashingHashJoinBlockTransform::appendBlock(Block block)
+{
+    handleOverLimitBlock();
```

Comment:
`handleOverLimitBlock()` can be moved to `getFinalOutputBlock` so that the number of calls is much less.
```
Block SquashingHashJoinBlockTransform::getFinalOutputBlock()
{
    Block final_block;
    if (blocks.empty())
    {
        final_block = {};
    }
    else if (blocks.size() == 1)
    {
        final_block = blocks[0];
    }
    else
    {
        final_block = mergeBlocks(std::move(blocks));
    }
    reset();
    handleOverLimitBlock()
    return final_block;
}
```

Reply:
done